### ISSUE TITLE
Automating JSON Feeds

### ISSUE THREAD

**Issue by tech2serve** [code4recovery/12-step-meeting-list#42]
A [request was made](https://wordpress.org/support/topic/automating-json-feeds/) in the support forums to have the plugin automatically update JSON feeds coming into the parent/top level site.

**Comment by wokat** [comment]
Hi all.

I've started work on this, as I did not find a matching branch:
https://github.com/aa-berlin/12-step-meeting-list/tree/add-cron-tasks

The state right now is:

- Little testing, but seems to work
- Naive refactoring that simply has the cron task do what the "Refresh"-Button does: Delete every meeting, reimport
- Cron task currently updates until it has run a certain hard-coded percentage of `max_execution_time`, but could probably just extend runtime?
- Added a locking mechanism; lock-breaking automatic based on start of execution and assumed max. possible run time of originator (assumption: has same predefined runtime as current process)
- Have no clue where to define "functions", so tried to leave unchanged code in place for sake of the history

I would suggest that we (or "I" :joy:) finish the following tasks before a pull-requests:

- Make schedule of import cron configurable
- Reuse local post-ids to allow for relating to posts, so update existing post of a meeting

**Comment by tech2serve** [comment]
Hi @wokat! Great to see you here. @brianw-area24 in particular had some ideas on this and I'd suggest you two connect. Probably the best way is for you to join our [Slack channel](https://join.slack.com/t/code4recovery/shared_invite/zt-ci34rbcx-r8ROFpwIqL6AYYmGZFp5xw).  He is pretty good about responding in there. A couple of others may chime in as well.

**Comment by wokat** [comment]
Current state (see PR above): 

* Cron jobs work! And faster than with full deletes/inserts before
* I would like to make the interval configurable still
* Updates are faster than before, but still slow; try adding some indexes and provide option to create them for the user
* No log viewing facilities yet (dump the html into a file like the meeting cache json?)
* Add a Break-Lock-Button? Import process is blocked based on start time and assumed max. duration derived from max_execution_time
* The Ajax-Version from before could poll the backend; now it just says "import running already" and stops without progress info
* Import is checking fields that seem to not be exported at the moment... Forgot which, though :sweat_smile:, or export might be stricter with contact fields than the meeting list display? Something like that...

Should we drop/postpone any of this for a first release?

**Comment by wokat** [comment]
Hi @tech2serve,

I updated my PR: https://github.com/code4recovery/12-step-meeting-list/pull/66

The schedule interval is configurable now. I also added a "Parent Region" per each feed, as CER of UK already will need that.

The other suggestions in my comment above I didn't touch much, though.

Let's see how you like this feature set so far.

**Comment by RECLO-CER** [comment]
Hi all, thanks for your service and attention to this issue. Would be really great to have auto syncing of JSON feeds between lists as currently we need to refresh our 5x linked JSON feeds manually once every couple of weeks or so which also takes a bit of time and also requires rearrangement of regions for some lists. Auto syncing, say once a day, would also help to ensure that the latest information is displayed on our site promptly. Apologies that I can't be of any more help technically as I am not a coder, but I definitely appreciate all your efforts and service! Matt RECLO, CER

**Comment by jcookson-infoblox** [comment]
Am looking at the "Import & Settings" page for the plugin and not seeing anything for automating import of feeds.  Where is this located?

**Comment by jcookson-infoblox** [comment]
Is there any chance of someone within the current dev-org reviving and starting to integrate the PR mentioned above (https://github.com/code4recovery/12-step-meeting-list/pull/66) into the app.  This item would allow for us to automate updates to meeting data.

**Comment by tech2serve** [comment]
Regrettably, the original person who did this has disappeared, and at this time none of the current C4R members have had the time to take it on.

The biggest thing needed, which anyone can do, is to get two TSML instances set up (in Docker containers would work), and run controlled tests to determine where it doesn't work. Or why it works sometimes (obviously @wokat had it working or he wouldn't have done a PR) and document steps to install it.

If you did want to take then on, I could probably work to update the base merge point of this PR. It has diverged quite a bit from our current `develop` branch.

**Comment by kiyote33** [comment]
I'm interested in the progress on this.  Willing to help, if that's needed to get it out....

**Comment by jcookson-infoblox** [comment]
As far as am aware, nobody has taken this up since the previous person worked on it.  If you are able to cherry-pick their code off their most recent commit & then merge it into todays code ... and get it working - would love to see this piece in the wild.

**Comment by kiyote33** [comment]
I've got this code running in a test environment where it's been executed 6 times, with 6 different results when refreshing the original 295 records (272, 590, 320, 295, 280, 601). Only once did I get the expected result (295). I am thinking maybe a simpler approach might work a little better:
1) Replace "Parent Region dropdown" with option to maintain the status quo (refresh manually) or once daily after midnight. 
2) Cron job would only fire if user admin has selected the option to run once daily.
3) The cron job script would spin through the json feed checking the LastUpdate field for a meeting update after the last feed update. When true, exit and execute code hooked on the Refresh button. When false, update the feed Last Update field and terminate.

Unless there's something obvious that I'm missing, I can give the above a go.  Anyone else have thoughts on this?

**Comment by kiyote33** [comment]
Interesting amendment to above testing.  A little patience and multiple passes is giving expected results regularly. The screen refresh is intermittently accurate. So far testing is actually going well with automated feed picking up changes accurately.

**Comment by 97eighty** [comment]
@kiyote33 
That’s great. I’m open for testing. Which branch?

**Comment by kiyote33** [comment]
Mornin 97,

 

I’m working with the code form #66. I haven’t touched the code base other than to take a quick look at it.

**Comment by kiyote33** [comment]
Weekend spent running this 3.7.2 branch with varying results; mostly bad.  Some success when only working with 1 feed but had no  wins after introducing a 2nd feed. Trying to get the 1st feed to work again was difficult. 

I would like to take a run at this with the following parameters: 
1) Retain "Parent Region dropdown" for setting parent of each feed.
2) Add option to create separate cron job for each feed (or not).
3) Keep cron job scheduling simple: each job should only execute once daily (nightly...).
4) Validate whether any feed records have changed since last processed and stop processing when no changes detected.
5)  Return to doing what the "Refresh" button does: When there are changes we Delete every meeting and then reimport.

When doing the above manually, I had a high success rate. I would expect similar results when automated.
Thoughts??? 
I may not be hitting the same thresholds as others were running. I was running a range of tests from 250 to 1000 meeting records. Most of my good results happened on the smaller datasets.

**Comment by jcookson-infoblox** [comment]
Responding to all comments from over the weekend, separating email comments via ">> next comment >>".  My comments start with “~~”.

>> initial comment >>

I've got this code running in a test environment where it's been executed 6 times, with 6 different results when refreshing the original 295 records (272, 590, 320, 295, 280, 601). Only once did I get the expected result (295). I am thinking maybe a simpler approach might work a little better:

~~ Something to think on, am the area webservant responsible for updating the feeds o our site.  Our feeds come in various size categories (# of feeds): 0-50 (2), 50-100 (1), 100-500 (1), 500-1000 (1), 1000+ (1) ~~

Replace "Parent Region dropdown" with option to maintain the status quo (refresh manually) or once daily after midnight.

~~ Seems fine, as long as the feeds are run concurrently ~~

Cron job would only fire if user admin has selected the option to run once daily.

The cron job script would spin through the json feed checking the LastUpdate field for a meeting update after the last feed update. When true, exit and execute code hooked on the Refresh button. When false, update the feed Last Update field and terminate.

Unless there's something obvious that I'm missing, I can give the above a go. Anyone else have thoughts on this?

>> next comment >>

Interesting amendment to above testing. A little patience and multiple passes is giving expected results regularly. The screen refresh is intermittently accurate. So far testing is actually going well with automated feed picking up changes accurately.

~~ encouraging ~~

>> next comment >>

Weekend spent running this 3.7.2 branch with varying results; mostly bad. Some success when only working with 1 feed but had no wins after introducing a 2nd feed. Trying to get the 1st feed to work again was difficult.

~~ interesting, when i try to refresh my feeds using Sel FF Plugin, it lags behind the UI and times out : wondering if you are having similar issue ~~

I would like to take a run at this with the following parameters:

Retain "Parent Region dropdown" for setting parent of each feed.
Add option to create separate cron job for each feed (or not).

~~ nice ... very nice ~~

Keep cron job scheduling simple: each job should only execute once daily (nightly...).

~~ my advice on this would be to schedule every cron to go off 30 mins after the previous one, to give time for larger feeds to go through the system ~~

Validate whether any feed records have changed since last processed and stop processing when no changes detected.

Return to doing what the "Refresh" button does: When there are changes we Delete every meeting and then reimport.
When doing the above manually, I had a high success rate. I would expect similar results when automated.

~~ auto-replace is good, the data source is responsible for the quality of their data ~~

Thoughts???
I may not be hitting the same thresholds as others were running. I was running a range of tests from 250 to 1000 meeting records. Most of my good results happened on the smaller datasets.

~~ from a test perspective, would suggest this range for # of records (what it represents),
0-50 : very small intergroup (small town,region)
51-100 : small intergroup (small town,region)
101-500 : medium intergroup (small city,region)
501-1000 : large intergroup (medium city / county)
1001-5000 : intergroup (large city / medium sized county)
5001+ : intergroup's/district's/area's (large city / metro-area [multi-counties])

James Cookson
Infoblox | Software Engineer II
m: 1.253.227.6336<tel:+12532276336>
***@***.******@***.***> | www.infoblox.com<http://www.infoblox.com>
[A picture containing text, clipart  Description automatically generated]

From: kiyote33 ***@***.***>
Reply-To: code4recovery/12-step-meeting-list ***@***.***>
Date: Sunday, June 20, 2021 at 10:41 PM
To: code4recovery/12-step-meeting-list ***@***.***>
Cc: James Cookson ***@***.***>, Comment ***@***.***>
Subject: Re: [code4recovery/12-step-meeting-list] Automating JSON Feeds (#42)


Weekend spent running this 3.7.2 branch with varying results; mostly bad. Some success when only working with 1 feed but had no wins after introducing a 2nd feed. Trying to get the 1st feed to work again was difficult.

I would like to take a run at this with the following parameters:

  1.  Retain "Parent Region dropdown" for setting parent of each feed.
  2.  Add option to create separate cron job for each feed (or not).
  3.  Keep cron job scheduling simple: each job should only execute once daily (nightly...).
  4.  Validate whether any feed records have changed since last processed and stop processing when no changes detected.
  5.  Return to doing what the "Refresh" button does: When there are changes we Delete every meeting and then reimport.

When doing the above manually, I had a high success rate. I would expect similar results when automated.
Thoughts???
I may not be hitting the same thresholds as others were running. I was running a range of tests from 250 to 1000 meeting records. Most of my good results happened on the smaller datasets.

**Comment by jcookson-infoblox** [comment]
re: "Are you talking about csv input? Or JSON feeds from other sites?" ... both, if it can orphan data, then it should be a concern.

**Comment by kiyote33** [comment]
Pictures worth a thousand words, so here's my proposal:
![AutomatingJsonFeeds](https://user-images.githubusercontent.com/57125427/123118446-e1011d80-d3ff-11eb-8eac-6b6ae25fd65e.png)

**Comment by jcookson-infoblox** [comment]
I love it .... looks simple and easy to configure.

**Comment by jcookson-infoblox** [comment]
is this UI on this branch ... so I could potentially pull it .. and look @ it ... you asked about scenarios / cases earlier and I have not forgotten about it, just been busy and the local heatwave has complicated things ....

**Comment by jcookson-infoblox** [comment]
also, please convey time slot choices or UI ... for setting up auto-refresh daily ...

**Comment by kiyote33** [comment]
Using the built-in wp_cron task scheduler comes with it's own limitations, especially when it comes to starting jobs at a specific time.  (i.e. the scheduler relies on the web server being polled or visited to activate scheduled tasks). So, if a job is scheduled for 1:am but no one visits the site until 8:am, then the job would run after 8:am.  My thought is just to run jobs utilizing the inherent 'once daily' schedule which should fire at the first opportunity after midnight. The jobs should run in succession, not concurrently.
I've pulled this branch, but haven't checked in any modifications yet.

**Comment by 97eighty** [comment]
Hi @kiyote33 

Can you create a branch based on the develop branch? Suggest naming kiyote/issue-42 or something similar so we can pull the branch and test. Thanks for your efforts.

**Comment by kiyote33** [comment]
I've spent a good chunk of time the last 3 weeks studying this issue, testing the existing code, and trying different approaches to getting it working. With no success there, I started from scratch and tried a minimal approach which ended with failure because of long running processes and complications with the ajax code.  So, rather than admit defeat completely I took a stab at setting each data source up with it's own wp-cron job, spinning through the data source to see if there were any records updated since the last import, and when found fire off a notification email to alert the website admins that the specific data source is now out of sync with what they are currently displaying. They can then take action to resolve the issue by refreshing the data source. To my surprise this has worked well without any issues so far.  Below are screen shots of what I have working. I can make the code available if others think this might be an approach we would like to continue forward with.
![DataSourceChangeDetection](https://user-images.githubusercontent.com/57125427/125023853-ef287e00-e03c-11eb-
![ChangesDetectedEmail](https://user-images.githubusercontent.com/57125427/125023893-023b4e00-e03d-11eb-891a-84eddddd14e0.png)
8eb2-39db3ee107ff.png)

**Comment by tech2serve** [comment]
Hey @kiyote33 

We appreciate you looking at this. I’ve been thinking lately about a very different approach, but one that has been done before for something similar.

Would you have some time next week for a Zoom call?  Can you shoot me an email to our devs@code4recovery.org address and perhaps we can chat about this and the other PR you worked on and opened?

Thanks,

Tim

**Comment by kiyote33** [comment]
Tim,
Your devs@code4recovery.org appears to have auto-response turned on. Are you monitoring that account?

**Comment by kiyote33** [comment]
Current state of issue #42: Above "Change Detection & Notify" code available for others to test at https://github.com/kiyote33/12-step-meeting-list-2.git. The new code is merged with the 3.11.2 code base.  Please give it a try and let me know what you think...

**Comment by district43aa** [comment]
Where do I find the JSON URL or the URL that I give to another aa website to have our meetings listed in their site. Are there any instructions or help out there please? 
🙏

**Comment by tech2serve** [comment]
You should navigate to your Import & Settings page, and then convert the Sharing to Open:

<img width="278" alt="image" src="https://user-images.githubusercontent.com/31904316/133895894-9aef2564-0a2e-43af-962c-4f4904030d48.png">

You will then be presented a link that you can share with another webmaster.

**Comment by district43aa** [comment]
My Data source does not have the option for auto update. How did you get that?

https://user-images.githubusercontent.com/57125427/123118446-e1011d80-d3ff-11eb-8eac-6b6ae25fd65e.png

**Comment by tech2serve** [comment]
It currently doesn't exists. That is what this issue is about. 

We've not yet been successful at automating this process. You'll still need to manually update as often as you wish (weekly?)
